File: proselint/checks/dates_times/am_pm.py
"""
a.m. / p.m.

---
layout:     post
source:     Garner's Modern American Usage
source_url: http://bit.ly/1T4alrY
title:      a.m. & p.m.
date:       2014-06-10 12:31:19
categories: writing
---

"""

from proselint.registry.checks import Check, engine, types

check_lowercase_periods = Check(
    check_type=types.ExistenceSimple(pattern=r"\d{1,2} ?[ap]m"),
    path="dates_times.am_pm.lowercase_periods",
    message="It's standard to use periods for lowercase 'a.m.' or 'p.m.'.",
    engine=engine.Fast(opts=engine.RegexOptions(case_insensitive=False)),
)

check_spacing = Check(
    check_type=types.ExistenceSimple(pattern=r"\d{1,2}[ap]\.?m\.?"),
    path="dates_times.am_pm.spacing",
    message="It's standard to put a space before 'a.m.' or 'p.m.'.",
)

check_midnight_noon = Check(
    check_type=types.ExistenceSimple(pattern=r"12 ?[ap]\.?m\.?"),
    path="dates_times.am_pm.midnight_noon",
    message="'12 a.m.' and '12 p.m.' are ambiguous. Use 'midnight' or 'noon'.",
)

check_redundancy = Check(
    check_type=types.Existence(
        items=(
            r"\d{1,2} ?a\.?m\.? in the morning",
            r"\d{1,2} ?p\.?m\.? in the evening",
            r"\d{1,2} ?p\.?m\.? at night",
            r"\d{1,2} ?p\.?m\.? in the afternoon",
        )
    ),
    path="dates_times.am_pm.midnight_noon",
    message="'a.m.' is always morning; 'p.m.' is always night.",
)

__register__ = (
    check_lowercase_periods,
    check_spacing,
    check_midnight_noon,
    check_redundancy,
)

File: proselint-rs/crates/proselint-checks/src/dates_times/am_pm.rs
use const_format::str_replace;

use proselint_registry::checks::{Check, CheckType, Padding};

pub const EXAMPLES_PASS: &[&str] = &[
    "Smoke phrase with nothing flagged.",
    "It happened at 7 a.m.",
    "On Wed, Sep 21, 2016 at 11:42 AM -0400, X wrote:",
    "It happened at 7 a.m.",
    "It happened at noon.",
    "It happened at 7 a.m.",
];
pub const EXAMPLES_FAIL: &[&str] = &[
    "It happened at 7 am.",
    "It happened at 7a.m.",
    "It happened at 12 a.m.",
    "It happened at 12 p.m.",
    "It happened at 7a.m. in the morning.",
];

const CHECK_LOWERCASE_PERIODS: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: str_replace!(Padding::WordsInText.as_str(), "{}", r"\d{1,2} ?[ap]m"),
		unicode: true,
		exceptions: &[],
	},
	path: "dates_times.am_pm.lowercase_periods",
	msg: "With lowercase letters, periods are standard.",
	ignore_case: false,
	..Check::default()
};

const CHECK_SPACING: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: r"\b\d{1,2}[ap]\.?m\.?",
		unicode: true,
		exceptions: &[],
	},
	path: "dates_times.am_pm.spacing",
	msg: "It's standard to put a space before 'a.m.' or 'p.m.'.",
	..Check::default()
};

const CHECK_MIDNIGHT_NOON: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: r"\b12 ?[ap]\.?m\.?",
		unicode: true,
		exceptions: &[], 
	},
	path: "dates_times.am_pm.midnight_noon",
    msg: "12 a.m. and 12 p.m. are wrong and confusing. Use 'midnight' or 'noon'.",
	..Check::default()
};

const CHECK_REDUNDANCY: Check = Check {
	check_type: CheckType::Existence {
		items: &[
            r"\b\d{1,2} ?a\.?m\.? in the morning",
            r"\b\d{1,2} ?p\.?m\.? in the evening",
            r"\b\d{1,2} ?p\.?m\.? at night",
            r"\b\d{1,2} ?p\.?m\.? in the afternoon",
		],
		unicode: true,
		padding: Padding::None,
		dotall: false,
		exceptions: &[],
	},
	path: "dates_times.am_pm.redundancy",
	msg: "'a.m.' is always morning; 'p.m.' is always night.",
	..Check::default()
};

pub const REGISTER: &[Check] = &[
	CHECK_LOWERCASE_PERIODS,
	CHECK_SPACING,
	CHECK_MIDNIGHT_NOON,
	CHECK_REDUNDANCY,
];
